Pluggability

iankronquist opened the pull request:
1. Make configs pluggable
2. Make back ends pluggable
3. Make spicedham a class

@willkg r?

iankronquist committed:
Remove old backend module

load_backend has been merged with spicedham/__init__.py

iankronquist committed:
Import load_backend from spicedham

The old spicedham.backend module has been removed

iankronquist committed:
Move load_backend to __init__

1. This way we load both types of files in one place
2. Soon I will wrap this all in a class which will get rid of that nagging
global variable.

iankronquist committed:
Change load_backend to fix bugs and simplify logic

	* I forgot to load the class for non-djangoorm backends (oops).
	* I realized that both branches could share logic.

iankronquist committed:
Modify tests to use the new backends system

iankronquist committed:
Add a list of backends plugins

iankronquist committed:
Make configs pluggable

iankronquist committed:
Remove config.py

Migrating it into the spicedham class.

iankronquist committed:
Make spicedham a class and bring in _load_config

iankronquist committed:
Convert plugins to use the Spicedham class

iankronquist committed:
Update tests to use Spicedham class

iankronquist committed:
Update test to use Spicedham tests

iankronquist committed:
Remove silly argument

iankronquist committed:
Update report to use Spicedham class

willkg commented on spicedham/__init__.py:
```diff
@@ -1,45 +1,74 @@
+    # Default config values
+    config = {
+        'backend': 'sqlalchemy',
+        'engine': 'sqlite:///:memory:',
```
I'm not sure this is a good default. With this default, it means that when training is done all the training data is lost.

It's probably better to do something like "sqlite:///./spicedham.db" or something like that.

willkg commented on spicedham/__init__.py:
```diff
@@ -1,45 +1,74 @@
+        'engine': 'sqlite:///:memory:',
+        'nonsensefilter': {'filter_match': 1, 'filter_miss': None},
+        'digitdestroyer': {'filter_match': 1, 'filter_miss': None},
+    }
```
Instead of putting plugin defaults here, it's probably better to put them in their respective classes.

willkg commented on spicedham/__init__.py:
```diff
@@ -1,45 +1,74 @@
-        return average_score / total
-    else:
-        return 0
+            self._classifier_plugins.append(pluginClass(self))
```
One ramification of this infrastructure is that it requires Python libraries to be installed in order to enable plugins. With fjord, we're using `vendor/` so we don't actually "install" anything. I'm not entirely sure whether this will work or not.

willkg commented on spicedham/__init__.py:
```diff
@@ -1,45 +1,74 @@
+                name=self.config['backend'])
+            entry_point = next(entry_point)
+        except StopIteration:
+            raise NoBackendFoundError
```
This should be an instance--not a class. e.g.:

```
raise NoBackendFoundError()
```

willkg replied:
Oops--I didn't know this about Python. Turns out if there are no args, you can raise an instance and that's ok. My bad!

willkg commented on spicedham/__init__.py:
```diff
@@ -1,45 +1,74 @@
+            for key in config_plugin_obj.keys():
+                config[key] = config_plugin_obj[key]
+
+    def train(self, training_data, is_spam):
```
What is "training_data" and "is_spam" here?

Does "is_spam" reduce the usefulness of this classifier to just spam? Maybe "is_bad" is a better argument implying this is the "bad side" of a set?

iankronquist replied:
I think I explain what these are in the README docs (although those need to be updated). Please tell me if they are unclear. I will change 'is_spam' to 'is_bad' though

willkg replied:
This should get a doc-string and then the doc-string can get extracted into API documentation. That's a lot easier than maintaining external docs in the README plus it means the documentation is in the code alongside the thing it's documenting.

willkg commented on spicedham/__init__.py:
```diff
@@ -1,45 +1,74 @@
+        for plugin in self._classifier_plugins:
+            plugin.train(training_data, is_spam)
+
+    def classify(self, classification_data):
```
What is "classification_data" here? What shape is it? What's an example?

iankronquist replied:
See above comment.

willkg commented on spicedham/__init__.py:
```diff
@@ -1,45 +1,74 @@
+                average_score += value
+        # On rare occasions no _plugins will give scores. If so, return 0
+        if total > 0:
+            return average_score / total
```
Is this integer or float division?

willkg commented on spicedham/baseplugin.py:
```diff
@@ -2,6 +2,8 @@ class BasePlugin(object):
     """
     A base class for spicedham plugins.
     """
+    def __init__(self, spiced_ham_object):
```
Does this imply there's a circular reference between SpicedHam and the backend?

willkg replied:
Seems like instead of passing in a SpicedHam instance, it's better to pass in the config and backend objects.

iankronquist replied:
New commits take care of everything but this.

iankronquist committed:
Change default config options

willkg commented on tests/test_bayes.py:
```diff
@@ -24,15 +24,13 @@ def _training(self, bayes):
         for message, is_spam in izip(messagePairs, cycle((True, False))):
             bayes.train(message, is_spam)
 
```
I can't tell if I'm reading this wrong or not, but it seems like `test_classify` doesn't actually test classification at all here. Is that right? If so, it could use a FIXME or some other comment suggesting there's missing stuff here.

iankronquist replied:
I think it does test classification. On line 18 you can see that it asserts the score should be 0.5. If you think that this is not rigorous enough, I can change it.

willkg commented:
I had some issues, but mostly they're cosmetic and documentation related.

I'll have to see if the entrypoints thing will work with Fjord and the way we do dependencies. I'm not sure if it will. The way entrypoints works is that they get registered when you pip install something and that gets added to the package metadata. We're not pip installing anything, so I think the entry points don't get indexed anywhere.

iankronquist commented:
I hadn't considered that. That may be a major issue.

iankronquist committed:
Add sensible defaults when getting configuration

iankronquist committed:
Largely cosmetic changes

* Use 'is_bad' so it's clear that we can classify more than just spam
* Make sure it's floating point division and not integer division
    (It should be float division anyway, but future plugins may not respect
	this)
* Raise an instance not a class

willkg commented:
I did some poking around (but didn't actually test it) and I'm pretty sure entrypoints aren't going to work.

Instead, I think you should do what I did with dennis where I (ab)use the fact that every class knows its subclasses and use that as the list of possible options.

Here's the "all_subclasses" function:

https://github.com/willkg/dennis/blob/master/dennis/tools.py#L204

Here's a usage of it:

https://github.com/willkg/dennis/blob/master/dennis/translator.py#L647

This doesn't work unless the things that subclass BasePlugin and friends have been imported _before_ the SpicedHam object gets created. With fjord, that'll work fine because we load everything on Earth before anything starts doing stuff. In Dennis' case, that doesn't happen since it's a command line utility. So Dennis needs to grow an entrypoints thing in order for plugins to work at all.

iankronquist committed:
Add docstrings for API, change `is_bad` to `match`

A positive match may be bad, as is the case for spam, or good, as is the case
for usefulness.

iankronquist committed:
Handle case when configs for plugins don't exist

iankronquist committed:
Pass config and backend to plugins

They should have references to the config and the backend, but not to
the whole Spicedham object.

iankronquist committed:
Update tests to pass a config and a backend

iankronquist committed:
Instatiate plugins with a config and a backend

iankronquist commented:
@willkg all comments addressed. I'll go take a look at dennis.

iankronquist committed:
A file to import for the side effects

Gross.

iankronquist committed:
Change from entry points to subclass lookup

Recursively look up subclasses. Note that we also import a file for import
side effects. That's gross.

iankronquist committed:
Add a base class for config objects

Subclasses dict so we can use subclass lookup. Not optimal but a workable
solution.

iankronquist committed:
flake8 cleanup

iankronquist commented:
@willkg removed entry points

iankronquist committed:
Don't subclass dict, make a load_conf class

Instead of relying on a class to have all the properties of a dict,
have it return a real one.
This needs to be a class and not just a function so we can inspect it using
__subclasses__

iankronquist committed:
Remove extraneous and problematic import

iankronquist committed:
Pass in config object instead of importing one

iankronquist committed:
Pass in config for use by subclasses

iankronquist committed:
Pass config dict to backends on instantiation

iankronquist committed:
Refactor BaseWrapper to BaseBackend

willkg commented on spicedham/__init__.py:
```diff
@@ -1,45 +1,87 @@
+from baseconfig import BaseConfig
+# Import for the side effect of getting access to subclasses
+# TODO: This is gross
+import gottaimportthemall
```
These should be relative imports: http://legacy.python.org/dev/peps/pep-0328/#rationale-for-relative-imports

willkg replied:
Also, "gottaimportthemall" is not in the spicedham module. It should be moved.

willkg commented on spicedham/__init__.py:
```diff
@@ -1,45 +1,87 @@
+            subc.extend(self.all_subclasses(d))
+        return subc
+
+    def __init__(self):
```
It'd be easier coupling if this took a configuration object as an argument. Then it doesn't have to deal with loading it which has some order-dependent edge cases.

willkg commented on spicedham/__init__.py:
```diff
@@ -1,45 +1,87 @@
+        for config_plugin_obj in self.all_subclasses(BaseConfig):
+            confer = config_plugin_obj().load_config()
+            for key in confer.keys():
+                self.config[key] = confer[key]
```
I don't think you want it to try to load all possible configuration objects. What you want is for this to load the configuration object specified by the user. Then it pulls the configuration from that.

Otherwise you end up in order-dependent edge cases.

iankronquist replied:
How can the user specify what the first config will be? Maybe we could put that in a config... oh wait.
Maybe I could call `sorted` on that list and select the first one so it will be a bit more deterministic?

willkg commented on spicedham/basebackend.py:
```diff
@@ -1,16 +1,18 @@
 from itertools import izip, repeat
```
I'd rename this file to just "backend.py" since it holds all the backend infrastructure that's required.

willkg commented on spicedham/baseconfig.py:
```diff
@@ -0,0 +1,3 @@
+class BaseConfig(object):
```
I'd rename this to just "config.py" and have it hold all the configuration infrastructure.

willkg commented on spicedham/baseplugin.py:
```diff
@@ -2,6 +2,8 @@ class BasePlugin(object):
     """
```
I'd rename this to just "plugin.py" and have it hold all the plugin infrastructure.

iankronquist committed:
Move hackish import inside the library

iankronquist committed:
Make imports absolute

iankronquist committed:
Pass in a config object, don't guess

Take a config object as an optional argument and resist the temptation to guess

iankronquist committed:
Refactor - remove 'base' from file names

iankronquist commented:
@willkg changes made.

willkg commented on spicedham/__init__.py:
```diff
@@ -1,45 +1,82 @@
+        Load config, backend, and plugins
+        """
+        if config is not None:
+            self.config = config
```
I'm not sure I understand why it could be None here. Seems like a bad thing to have no configuration.

Also the docstring is now out of date.

willkg replied:
Oh, I see--it'll use the class level default config. That's fine. The docstring should still get fixed.

willkg commented:
Fix the docstring and then this is good enough to land.

iankronquist committed:
Fix docstring

iankronquist commented:
@willkg fixed docstring.

willkg commented:
Cool! r+
